File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/controller/overlay/handler/PolygonOverlayHandler.kt
package dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.handler

import com.naver.maps.map.overlay.Overlay
import com.naver.maps.map.overlay.PolygonOverlay
import dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.OverlayHandler
import dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.OverlayHandler.Companion.getterName
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay.NPolygonOverlay
import io.flutter.plugin.common.MethodChannel

internal interface PolygonOverlayHandler : OverlayHandler {
    fun handlePolygonOverlay(
        polygonOverlay: Overlay, method: String, arg: Any?, result: MethodChannel.Result,
    ) = (polygonOverlay as PolygonOverlay).let { p ->
        when (method) {
            NPolygonOverlay.coordsName -> setCoords(p, arg!!)
            NPolygonOverlay.colorName -> setColor(p, arg!!)
            NPolygonOverlay.holesName -> setHoles(p, arg!!)
            NPolygonOverlay.outlineColorName -> setOutlineColor(p, arg!!)
            NPolygonOverlay.outlineWidthName -> setOutlineWidth(p, arg!!)
            NPolygonOverlay.outlinePatternName -> setOutlinePattern(p, arg!!)
            getterName(NPolygonOverlay.boundsName) -> getBounds(p, result::success)
            else -> result.notImplemented()
        }
    }

    fun setCoords(polygonOverlay: PolygonOverlay, rawCoords: Any)

    fun setColor(polygonOverlay: PolygonOverlay, rawColor: Any)

    fun setHoles(polygonOverlay: PolygonOverlay, rawHoles: Any)

    fun setOutlineColor(polygonOverlay: PolygonOverlay, rawColor: Any)

    fun setOutlineWidth(polygonOverlay: PolygonOverlay, rawWidthDp: Any)

    fun setOutlinePattern(polygonOverlay: PolygonOverlay, rawPatternDpList: Any)

    fun getBounds(polygonOverlay: PolygonOverlay, success: (bounds: Map<String, Any>) -> Unit)
}

File: ios/flutter_naver_map/Sources/flutter_naver_map/controller/overlay/handler/PolygonOverlayHandler.swift
import NMapsMap
import Flutter

internal protocol PolygonOverlayHandler: OverlayHandler {
    func setCoords(_ polygonOverlay: NMFPolygonOverlay, rawCoords: Any)

    func setColor(_ polygonOverlay: NMFPolygonOverlay, rawColor: Any)

    func setHoles(_ polygonOverlay: NMFPolygonOverlay, rawHoles: Any)

    func setOutlineColor(_ polygonOverlay: NMFPolygonOverlay, rawColor: Any)

    func setOutlineWidth(_ polygonOverlay: NMFPolygonOverlay, rawWidthDp: Any)
    
    func setOutlinePattern(_ polygonOverlay: NMFPolygonOverlay, patternList: Any)

    func getBounds(_ polygonOverlay: NMFPolygonOverlay, success: (_ bounds: Dictionary<String, Any>) -> Void)
}

internal extension  PolygonOverlayHandler {
    func handlePolygonOverlay(polygonOverlay: NMFOverlay,
                              method: String,
                              args: Any?,
                              result: @escaping FlutterResult) {
        let polygonOverlay = polygonOverlay as! NMFPolygonOverlay
        switch method {
        case NPolygonOverlay.coordsName: setCoords(polygonOverlay, rawCoords: args!)
        case NPolygonOverlay.colorName: setColor(polygonOverlay, rawColor: args!)
        case NPolygonOverlay.holesName: setHoles(polygonOverlay, rawHoles: args!)
        case NPolygonOverlay.outlineColorName: setOutlineColor(polygonOverlay, rawColor: args!)
        case NPolygonOverlay.outlineWidthName: setOutlineWidth(polygonOverlay, rawWidthDp: args!)
        case NPolygonOverlay.outlinePatternName:
            setOutlinePattern(polygonOverlay, patternList: args!)
        case getterName(NPolygonOverlay.boundsName): getBounds(polygonOverlay, success: result)
        default: result(FlutterMethodNotImplemented)
        }
    }
}
